### PR TITLE
8269088: C2 fails with assert(!n->is_Store() && !n->is_LoadStore()) failed: no node with a side effect

### PR DESCRIPTION
**Pull request by rwestrel**
In PhaseIdealLoop::try_sink_out_of_loop(), a Store is not expected to
be a candidate for sinking: if a Store is in a loop, there has to be
at least a memory Phi keeping the Store in the loop. In the case of
the failure, a Store has control set in a loop but the Store is not
connected to a loop memory Phi. The reason for that is that the Store
was moved out of loop by PhaseIdealLoop::try_move_store_after_loop()
which moved the Store to the lca of its uses (other than the loop
Phi), in this case another loop.

It doesn't look right that a Store is removed from a loop only to be
added to another one so I propose fixing that by placing the Store
right out of the loop it's being moved out of (similar to
PhaseIdealLoop::try_sink_out_of_loop()) rather than at the lca of
uses.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269088](https://bugs.openjdk.java.net/browse/JDK-8269088): C2 fails with assert(!n->is_Store() && !n->is_LoadStore()) failed: no node with a side effect

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to dcb1e6467a539aee184d751717a0a0bdf0f3a541
 * [Igor Veresov](https://openjdk.java.net/census#iveresov) (@veresov - **Reviewer**) ⚠️ Review applies to dcb1e6467a539aee184d751717a0a0bdf0f3a541
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/133/head:pull/133` \
`$ git checkout pull/133`

Update a local copy of the PR: \
`$ git checkout pull/133` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/133/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 133`

View PR using the GUI difftool: \
`$ git pr show -t 133`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/133.diff">https://git.openjdk.java.net/jdk17/pull/133.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopopts.cpp [hunk 1]
```diff
@@ -949,6 +949,9 @@ void PhaseIdealLoop::try_move_store_after_loop(Node* n) {
               assert(get_loop(lca) == outer_loop, "safepoint in outer loop consume all memory state");
             }
 #endif
+            lca = place_outside_loop(lca, n_loop);
+            assert(!n_loop->is_member(get_loop(lca)), "control must not be back in of loop");
```
Comment (neliasso) [inline src/hotspot/share/opto/loopopts.cpp:953]: assert comment needs adjusting

**Review by vnkozlov — APPROVED** [review]
@neliasso please run mach5 testing for this.

**Review by veresov — APPROVED** [review]

File: src/hotspot/share/opto/loopopts.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopopts.cpp]: Thanks for the review. Fixed.

**Review by neliasso — APPROVED** [review]

**Comment by rwestrel** [comment]
@vnkozlov @veresov @neliasso thanks for the reviews

**Pull request closed without merging**